salt/tests: Add unit tests for `metalk8s_drain`

gdemonet opened the pull request:
This PR adds unit tests providing 100% coverage on the `metalk8s_drain` module.

To do so, it introduces:
- some utilities for capturing and checking logs
- an API mock which allows the tester to manage responses and objects
directly, while the tested code can interact with it through
`metalk8s_kubernetes` methods
- a mock for the `time` module, which allows the drain process to run
faster than real time, and also controls the API mock through timed
events
- a mock for the eviction API

See: #2266

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/2266-metalk8s-drain-tests` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `improvement/2266-metalk8s-drain-tests`
* Rebase `improvement/2266-metalk8s-drain-tests` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

TeddyAndrieux commented on salt/_modules/metalk8s_drain.py:
```diff
@@ -248,8 +249,11 @@ def get_pod_controller(self, pod):
             pod['metadata']['namespace'], controller_ref
         )
         if not response:
+            meta = pod['metadata']
```
You do not use this variable, so either remove it or use it in the `format`

gdemonet replied:
Done [`3a23fec` (#2700)](https://github.com/scality/metalk8s/pull/2700/commits/3a23fec5f2bb63f3c1442e25a6264337db33f7da)

TeddyAndrieux commented on salt/tests/unit/modules/test_metalk8s_drain.py:
```diff
@@ -0,0 +1,347 @@
+
+    def test_virtual_nominal(self):
+        """Nominal behaviour for `__virtual__`."""
+
```
```suggestion
```

gdemonet replied:
Done [`c43efc1` (#2700)](https://github.com/scality/metalk8s/pull/2700/commits/c43efc13e3cfc777e053f928bd0de1a8b6d4b70f)

TeddyAndrieux commented on salt/tests/unit/modules/test_metalk8s_drain.py:
```diff
@@ -0,0 +1,347 @@
+    ])
+    def test_virtual_fail_import(self, _, package):
+        """Behaviour for `__virtual__` on failed imports."""
+
```
```suggestion
```

gdemonet replied:
Done [`c43efc1` (#2700)](https://github.com/scality/metalk8s/pull/2700/commits/c43efc13e3cfc777e053f928bd0de1a8b6d4b70f)

TeddyAndrieux reviewed (changes requested):
Ok looks good to me, just this PR should target `development/2.4` and not 2.6

Also to me would make sense to split at least the change that goes in `metalk8s_drain.py` and the tests (and maybe also split tests utils from the rest)

gdemonet committed:
salt: Small bugfixes in metalk8s_drain module

Some error cases (discovered when adding unit tests) were not properly
handled given how we pass plain dicts for K8s objects (and not instances
from the `kubernetes` client library).

gdemonet committed:
salt/tests: Add utility for parameterize from YAML

There was a common pattern appearing in other unit test modules for
extracting test cases from a YAML listing.
This helper allows to simply pass the list of test cases to a basic
decorator.

See: #2266

gdemonet committed:
salt/tests: Define custom Kubernetes API mock

This mock allows to perform simple CRUD operations on an in-memory
database following the same principles as K8s API, and is built to be
used through `metalk8s_kubernetes` mocked methods (when testing other
execution modules relying on them).
In addition, this mock can be configured to evolve through "time", using
a list of "events", which are particularly useful to emulate the
reconciliation loops happening in a K8s cluster, while running our tests
much faster than real-time.

See: #2266

gdemonet committed:
salt/tests: Add utilities for capturing logs

This will allow testers of Salt modules to easily capture logs from the
tested module and check if they match what's expected.

See: #2266

gdemonet committed:
salt/tests: Add unit tests for `metalk8s_drain`

This module is slightly different from others, as most logic lives
within the `Drain` class, which is responsible for performing
side-effects. Refactoring this to provide better isolation would be too
long, so we test it as a black box and simply mock K8s API (and the
time).

In the YAML test cases, lists of Pods, ReplicaSets, etc. are stored
under the `datasets` key, which is then read by the mocks when setting
up a test case.

See: #2266

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/2266-metalk8s-drain-tests` with contents from `improvement/2266-metalk8s-drain-tests`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/2266-metalk8s-drain-tests origin/development/2.5
 $ git merge origin/improvement/2266-metalk8s-drain-tests
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/2266-metalk8s-drain-tests
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2266-metalk8s-drain-tests` into
`development/2.4`
* [w/2.5/improvement/2266-metalk8s-drain-tests](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/2266-metalk8s-drain-tests) will be merged into `development/2.5` 
* [w/2.6/improvement/2266-metalk8s-drain-tests](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/2266-metalk8s-drain-tests) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

gdemonet commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

gdemonet commented:
/wait

bert-e commented:
# History mismatch



Merge commit #89b0497f3fb6529b7a7dc14126707d6b7ee82ad2 on the integration branch
`w/2.5/improvement/2266-metalk8s-drain-tests` is merging a branch which is neither the current
branch `improvement/2266-metalk8s-drain-tests` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/2266-metalk8s-drain-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
@TeddyAndrieux @alexandre-allard-scality commits were squashed!

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/2266-metalk8s-drain-tests` with contents from `improvement/2266-metalk8s-drain-tests`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/2266-metalk8s-drain-tests origin/development/2.5
 $ git merge origin/improvement/2266-metalk8s-drain-tests
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/2266-metalk8s-drain-tests
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2266-metalk8s-drain-tests` into
`development/2.4`
* [w/2.5/improvement/2266-metalk8s-drain-tests](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/2266-metalk8s-drain-tests) will be merged into `development/2.5` 
* [w/2.6/improvement/2266-metalk8s-drain-tests](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/2266-metalk8s-drain-tests) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

gdemonet commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/999da80d0da05c8d9d50ec27e63eae50028103ea) did not succeed in branch w/2.6/improvement/2266-metalk8s-drain-tests.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
